Smooth Line View/SmoothLineView.m

```diff
@@ -179,63 +187,40 @@ - (void) touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
         
         // initializes our point records to current location
         self.previousPoint = [touch previousLocationInView:self];
-        self.previousPreviousPoint = [touch previousLocationInView:self];
         self.currentPoint = [touch locationInView:self];
         
         CGPathMoveToPoint(_path, NULL, self.currentPoint.x, self.currentPoint.y);
     }
 }
 
-- (void) touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
+- (void) panGestureCaught:(UIPanGestureRecognizer*)sender
 {
-    if (event.allTouches.count == 1)
-    {
-        UITouch *touch = [touches anyObject];
-        
-        /*
-         CGPoint point = [touch locationInView:self];
-         
-         // if the finger has moved less than the min dist ...
-         CGFloat dx = point.x - self.currentPoint.x;
-         CGFloat dy = point.y - self.currentPoint.y;
-         
-         if ((dx * dx + dy * dy) < kPointMinDistanceSquared) {
-         // ... then ignore this movement
-         return;
-         }
-         */
-        
-        // update points: previousPrevious -> mid1 -> previous -> mid2 -> current
-        self.previousPreviousPoint = self.previousPoint;
-        self.previousPoint = [touch previousLocationInView:self];
-        self.currentPoint = [touch locationInView:self];
-        
-        CGPoint mid2 = midPoint(self.currentPoint, self.previousPoint);
-        
-        // to represent the finger movement, add a quadratic bezier path
-        // from current point to mid2, using previous as a control point
-        CGPathAddQuadCurveToPoint(_path, NULL,
-                                  self.previousPoint.x, self.previousPoint.y,
-                                  mid2.x, mid2.y);
-    }
-    else
-    {
-        [self setPath:self.bzPath];
-    }
+    LogMessage(DBG, @"panGestureCaught %@", sender);
+    
+    NSAssert(sender == self.panGesture, @"panGestureCaught from unknown sender");
+    
+    CGPoint currentPoint = [sender locationInView:self];
+    self.previousPoint = self.currentPoint;
+    self.currentPoint = currentPoint;
+    
+    CGPoint mid2 = midPoint(self.currentPoint, self.previousPoint);
+    
+    // to represent the finger movement, add a quadratic bezier path
+    // from current point to mid2, using previous as a control point
+    CGPathAddQuadCurveToPoint(_path, NULL,
+                              self.previousPoint.x, self.previousPoint.y,
+                              mid2.x, mid2.y);
     
     [self setNeedsDisplayInRect:CGPathGetBoundingBox(_path)];
 }
 
 - (void) touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
 {
-    if (event.allTouches.count == 1)
+    UIBezierPath* path = [UIBezierPath bezierPathWithCGPath:_path];
+    if (!path.empty && !CGPathEqualToPath(path.CGPath, self.bzPath.CGPath))
```

Comment:
I realized that simply tapping while in freehand mode added a pathsnapshot with a MoveTo command appended to the end, and each tap added another snapshot. So you would have to tap undo as many times as you tapped to then eventually be able to physically see an undo happen.

This check `!CGPathEqualToPath(path.CGPath, self.bzPath.CGPath)` was meant to make it such that we don't add snapshots if nothing has changed and take this into account, but the path is always different in this case because of that MoveTo command at the end. I haven't found a way yet to strip empty MoveTos to make this problem go away. Any insight here?

Reply:
Completely independent from panGesture, this is an existing issue.